br/pkg/kms/gcp.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 PingCAP, Inc. Licensed under Apache-2.0.
+
+package kms
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/encryptionpb"
+	"golang.org/x/oauth2/google"
+)
+
+const (
+	GcpKmsEndpoint       = "https://cloudkms.googleapis.com/v1/"
+	MethodDecrypt        = "decrypt"
+	KeyIdPattern         = `^projects/([^/]+)/locations/([^/]+)/keyRings/([^/]+)/cryptoKeys/([^/]+)/?$`
+	StorageVendorNameGcp = "gcp"
+)
+
+var KeyIdRegex = regexp.MustCompile(KeyIdPattern)
+
+type GcpKms struct {
+	config   *encryptionpb.MasterKeyKms
+	location string
+	client   *http.Client
+}
+
+func NewGcpKms(config *encryptionpb.MasterKeyKms) (*GcpKms, error) {
+	if config.GcpKms == nil {
+		return nil, errors.New("GCP config is missing")
+	}
+	if !KeyIdRegex.MatchString(config.KeyId) {
+		return nil, errors.Errorf("invalid key: '%s'", config.KeyId)
+	}
+	config.KeyId = strings.TrimSuffix(config.KeyId, "/")
+	location := strings.Join(strings.Split(config.KeyId, "/")[:4], "/")
+
+	client, err := google.DefaultClient(context.Background(), "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return nil, errors.Errorf("failed to create GCP client: %v", err)
+	}
+
+	return &GcpKms{
+		config:   config,
+		location: location,
+		client:   client,
+	}, nil
+}
+
+func (g *GcpKms) doJSONRequest(ctx context.Context, keyName, method string, data any) ([]byte, error) {
+	url := g.formatCallURL(keyName, method)
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to marshal request")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", url, strings.NewReader(string(jsonData)))
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to create request")
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	start := time.Now()
+	resp, err := g.client.Do(req)
+	if err != nil {
+		return nil, errors.Annotate(err, "request failed")
+	}
+	defer resp.Body.Close()
+
+	// TODO: Implement metrics
+	_ = time.Since(start)
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("request failed with status: %s", resp.Status)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to read response body")
+	}
+
+	return body, nil
+}
+
+func (g *GcpKms) formatCallURL(key, method string) string {
+	return fmt.Sprintf("%s%s/:%s?alt=json", GcpKmsEndpoint, key, method)
+}
+
+func (g *GcpKms) Name() string {
+	return StorageVendorNameGcp
+}
+
+func (g *GcpKms) DecryptDataKey(ctx context.Context, dataKey []byte) ([]byte, error) {
```

Comment:
Perhaps we can use the GCP SDK: https://pkg.go.dev/cloud.google.com/go/kms? That can be well-tested and easier to use.